### PR TITLE
Added Bucket Sort

### PR DESCRIPTION
**Pull request by Sanchita1304**
## Change Summary

- Written explained tutorial of Bucket Sort
- Added 2 Leet code problems as example in tutorial using different approaches.
- Attached 3 suggested problems

## Checklist

**If you haven't fulfilled the below requirements or even delete the entire checklist, your PR won't be reviewed and will be closed without notice. Regular contributors (with 10+ PRs) can skip this part.**

### General 

- [x] This Pull Request is all my own work. (You'll be blacklisted if you are caught for plagiarism.)
- [x] I've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md)
- [x] I've applied LaTex for all variables, formulas and time / space complexity instead of using backticks
- [x] I've started the app locally and verified all the content and all links (if applicable) are accessible correctly
- [x] I've included Complexity Analysis (Time Complexity & Space Complexity).
- [x] I've written my explanation well and it is easy to understand for beginners

### Tutorial

- [x] I've read and followed the [Tutorial Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#tutorial-template)
- [x] I've explained my topic well with 2 - 3 LC problems and no external problems are used. 
- [x] I've provided the full working solutions to the problems used in this tutorial.
- [x] I've provided suggested problems at the end with the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [ ] I've given credits / references if I use external resources. (For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).)

### Solutions

- [x] I've read and followed the [Solution Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#solution-template)
- [x] I've formatted my code well with [K&R Coding style](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)
- [x] I've confirmed that comments are put above each line rather than writing on the same line.
- [x] I've included a meaningful approach name for my solution. e.g. `## Approach 1: Two Pointers`.

### CODE REVIEW

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 1]
```diff
@@ -1,14 +1,316 @@
...
+    For each non-empty bucket, sort its contents using a sorting algorithm of your choice (e.g., insertion sort, quicksort, or another       bucket sort recursively).
+    
+4. Concatenation:
+    
+    Concatenate all the sorted buckets in order to obtain the final sorted array.
+    
+
+We can understand the Bucket Sort by using Scatter-Gather Approach.
+
+![bucket sort example](https://github.com/Sanchita1304/leetcode-the-hard-way/assets/90080911/7045afe4-e20e-4019-acbf-b13dfcb7b886)
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:46]: source? did you make it yourself?

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 2]
```diff
@@ -1,14 +1,316 @@
...
+- tutorial
+- sorting
+- algorithm
+- bucket sort
+
 ---
 
-<TutorialAuthors names="@TBC"/>
+<TutorialAuthors names="@Sanchita1304"/>
+<Contributors names="@wingkwong"/>
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:18]: i didn't contribute. please remove this line.

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 3]
```diff
@@ -1,14 +1,316 @@
...
+    
+    Iterate through the input array and place each element into the appropriate bucket based on its value. This is done by mapping the       element's value to the corresponding bucket.
+    
+3. Sort Each Bucket:
+    
+    For each non-empty bucket, sort its contents using a sorting algorithm of your choice (e.g., insertion sort, quicksort, or another       bucket sort recursively).
+    
+4. Concatenation:
+    
+    Concatenate all the sorted buckets in order to obtain the final sorted array.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:41]: the format could have been better. also don't know why there is some large spaces in 2 & 3.

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 4]
```diff
@@ -1,14 +1,316 @@
...
+    
+
+We can understand the Bucket Sort by using Scatter-Gather Approach.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:44]: what is Scatter-Gather? no explaination?

File: tutorials/basic-topics/sorting/bucket-sort.md (hunk 1 above)
Comment (Sanchita1304) [reply tutorials/basic-topics/sorting/bucket-sort.md:46]: Yes, I made it myself

File: tutorials/basic-topics/sorting/bucket-sort.md (hunk 2 above)
Comment (Sanchita1304) [reply tutorials/basic-topics/sorting/bucket-sort.md]: okk

File: tutorials/basic-topics/sorting/bucket-sort.md (hunk 3 above)
Comment (Sanchita1304) [reply tutorials/basic-topics/sorting/bucket-sort.md]: I have made changes to it and delete the spacing

File: tutorials/basic-topics/sorting/bucket-sort.md (hunk 4 above)
Comment (Sanchita1304) [reply tutorials/basic-topics/sorting/bucket-sort.md:43]: The image below it and the explanation is the scatter-gather approach

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 5]
```diff
@@ -1,14 +1,315 @@
...
+};
+
+```
+
+</TabItem>
+</Tabs>
+
+
+
+Bucket Sort is useful for sorting when the input data is uniformly distributed across a range of values.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:292]: put this at the end of Algorithm section.

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 6]
```diff
@@ -1,14 +1,315 @@
...
+
+    Divides the range of input values into a series of "buckets." The number of buckets is determined by the range between the minimum and maximum values plus one `(bucket_range = max_val - min_val + 1)`. Each bucket is initially empty and will be used to count how many times each value appears in the input array.
+
+3. Distribute Elements into Buckets: 
+
+    The algorithm iterates through the input array nums. For each element, it calculates which bucket it belongs to by subtracting the min_val. It then increments the count in that bucket. This step effectively counts the occurrences of each value in the input array.
+
+4. Gather Sorted Elements:
+
+     After counting the occurrences of each value in the buckets, the algorithm reconstructs the sorted array. It iterates through the buckets in order and, for each non-empty bucket, adds the corresponding value back to the sorted array as many times as indicated by the count in the bucket. This process ensures that the values are added to the sorted array in ascending order.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:85]: these texts are simply the translation of the code but they don't explain why. For example, beginners may ask why we need to subtract the min value and why we need to add it back later on. This part is not mentioned in above tutorial as well.

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 7]
```diff
@@ -1,14 +1,315 @@
...
+Bucket sort is a non-comparative sorting algorithm that works by distributing elements into a finite number of buckets, each representing a specific range of values. After distributing the elements, each bucket is sorted individually, typically using another sorting algorithm or recursively applying bucket sort. Finally, the sorted elements are concatenated to produce a fully sorted array. This algorithm is particularly useful when the input data is uniformly distributed across a range of values, making it efficient for certain types of data sets.
+
+## Algorithm
+
+1. Initialization:
+    
+    Create an array of empty buckets (containers), typically represented as lists or arrays.
+    Determine the range of input values and divide this range into equal-sized subranges, each corresponding to a bucket.
+    
+2. Distribution:
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:30]: the term should be consistent. either you can say distribute or scatter the elements.

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 8]
```diff
@@ -1,14 +1,315 @@
...
+     The algorithm begins by finding the minimum and maximum values in the input array nums. This is done using the `min_element` and `max_element` functions from the C++ Standard Library. These values are crucial because they determine the range of values that need to be sorted.
+
+2. Create Buckets: 
+
+    Divides the range of input values into a series of "buckets." The number of buckets is determined by the range between the minimum and maximum values plus one `(bucket_range = max_val - min_val + 1)`. Each bucket is initially empty and will be used to count how many times each value appears in the input array.
+
+3. Distribute Elements into Buckets: 
+
+    The algorithm iterates through the input array nums. For each element, it calculates which bucket it belongs to by subtracting the min_val. It then increments the count in that bucket. This step effectively counts the occurrences of each value in the input array.
+
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:82]: As you've mentioned

> For each non-empty bucket, sort its contents using a sorting algorithm of your choice (e.g., insertion sort, quicksort, or another bucket sort recursively).

people may ask where is the sort part here.

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 9]
```diff
@@ -1,14 +1,315 @@
...
+    Finally, the sorted array is returned as the result of the sortArray function.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    vector<int> sortArray(vector<int>& nums) {
+       // Find the minimum and maximum values in nums
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:98]: i would suggest to add the number as well. e.g. `// 1. Find the minimum and maximum values in nums`

(apply to all).

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 10]
```diff
@@ -1,14 +1,315 @@
...
+
+8. Return Sorted Array: 
+    After processing all buckets, the sorted_array will contain all elements from the input array nums sorted in ascending order. Return this sorted array as the result.
+
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:169]: similarly, add above 8 points in comment.

File: tutorials/basic-topics/sorting/bucket-sort.md [hunk 11]
```diff
@@ -1,14 +1,315 @@
...
+    }
+};
+
+```
+
+</TabItem>
+</Tabs>
+
+
+Here, "n" is the number of elements in the input array 'nums' and "k" is the range of values in the array.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/bucket-sort.md:210]: use `$n$` instead of quotes. (apply to all)

**Review by wingkwong — CHANGES_REQUESTED** [review]
feel free to reach me at [Discord](https://discordapp.com/users/495466997304262658) if you have any questions.

**Comment by wingkwong** [comment]
closing due to inactivity

**Pull request closed without merging**